rednose.py

```diff
@@ -33,367 +33,371 @@
-	def writeln(self, *a):
-		if self.__should_filter():
-			return
-		return self.__stream.writeln(*a)
-
-	# pass non-known methods through to self.__stream
-	def __getattr__(self, name):
-		if REDNOSE_DEBUG:
-			print("REDNOSE_DEBUG: getting attr %s" % (name,), file=sys.stderr)
-		return getattr(self.__stream, name)
+class RedNose(nose.plugins.Plugin):
+    env_opt = 'NOSE_REDNOSE'
+    env_opt_color = 'NOSE_REDNOSE_COLOR'
+    score = 199  # just under the `coverage` module
+
+    def __init__(self, *args):
+        super(RedNose, self).__init__(*args)
+        self.reports = []
+        self.error = self.success = self.failure = self.skip = 0
+        self.total = 0
+        self.stream = None
+        self.verbose = False
+        self.enabled = False
+        self.tree = False
+
+    def options(self, parser, env=os.environ):
+        global REDNOSE_DEBUG
+        rednose_on = bool(env.get(self.env_opt, False))
+        rednose_color = env.get(self.env_opt_color, 'auto')
+        REDNOSE_DEBUG = bool(env.get('REDNOSE_DEBUG', False))
+
+        parser.add_option(
+            "--rednose",
+            action="store_true",
+            default=rednose_on,
+            dest="rednose",
+            help="enable colour output (alternatively, set $%s=1)" % (self.env_opt,)
+        )
+        parser.add_option(
+            "--no-color",
+            action="store_false",
+            dest="rednose",
+            help="disable colour output"
+        )
+        parser.add_option(
+            "--force-color",
+            action="store_const",
+            dest='rednose_color',
+            default=rednose_color,
+            const='force',
+            help="force colour output when not using a TTY (alternatively, set $%s=force)" % (self.env_opt_color,)
+        )
+        parser.add_option(
+            "--immediate",
+            action="store_true",
+            default=False,
+            help="print errors and failures as they happen, as well as at the end"
+        )
+
+    def configure(self, options, conf):
+        if options.rednose:
+            self.enabled = True
+            termstyle_init = {
+                'force': termstyle.enable,
+                'off': termstyle.disable
+            }.get(options.rednose_color, termstyle.auto)
+            termstyle_init()
+
+            self.immediate = options.immediate
+            self.verbose = options.verbosity >= 2
+
+    def prepareTestResult(self, result):
+        """Required to prevent others from monkey patching the add methods."""
+        return result
+
+    def prepareTestRunner(self, runner):
+        runner.immediate = self.immediate
+
+        def _makeResult(self):
+            return ColourTextTestResult(self.stream, self.descriptions, self.verbosity, self.config, immediate=self.immediate)
+
+        import types
+        runner._makeResult = types.MethodType(_makeResult, runner)
+        return runner
+
+    def setOutputStream(self, stream):
+        self.stream = stream
```

Comment:
I can use the colorama thing I accidentally removed and set it to `self.stream`.